### PR TITLE
Reduce kolibri-server starting/stopping times

### LINKED ISSUE #51: more timeouts when starting kolibri-server
migrated issue from @benjaoming  https://github.com/learningequality/kolibri-server/issues/31#issuecomment-573701639

----

> I still reproduce this issue on Raspberry Pi. Investigating the timeout.
> It's not exactly the same issue, but pretty close. Starting doesn't time out, it's now stopping that times out :)

```
root@raspberrypi:~# apt upgrade
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Calculating upgrade... Done
The following package was automatically installed and is no longer required:
  rpi.gpio-common
Use 'apt autoremove' to remove it.
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
1 not fully installed or removed.
After this operation, 0 B of additional disk space will be used.
Do you want to continue? [Y/n] 
Setting up kolibri-server (0.3.5-0ubuntu1) ...
WARNING:root:No C Extensions available for this platform.

WARNING  Options file /home/pi/.kolibri/options.ini has been updated; server restart is required before change will take effect.
WARNING  Options file /home/pi/.kolibri/options.ini has been updated; server restart is required before change will take effect.
Job for kolibri-server.service failed because a timeout was exceeded.
See "systemctl status kolibri-server.service" and "journalctl -xe" for details.
invoke-rc.d: initscript kolibri-server, action "start" failed.
● kolibri-server.service - A high performance web server setup for Kolibri
   Loaded: loaded (/lib/systemd/system/kolibri-server.service; enabled; vendor preset: enabled)
   Active: failed (Result: timeout) since Mon 2020-01-13 14:52:55 GMT; 54ms ago
  Process: 8848 ExecStart=/etc/init.d/kolibri-server start (code=exited, status=0/SUCCESS)
  Process: 8972 ExecStop=/etc/init.d/kolibri-server stop (code=killed, signal=TERM)

Jan 13 14:52:17 raspberrypi kolibri-server[8848]: WARNING:root:No C Extensions available for this platform.
Jan 13 14:52:18 raspberrypi kolibri-server[8848]: WARNING  Options file /home/pi/.kolibri/options.ini has been updated; server restart is required before change will take effect.
Jan 13 14:52:19 raspberrypi kolibri-server[8848]: WARNING  Options file /home/pi/.kolibri/options.ini has been updated; server restart is required before change will take effect.
Jan 13 14:52:19 raspberrypi kolibri-server[8848]: debconf: DbDriver "passwords" warning: could not open /var/cache/debconf/passwords.dat: Permission denied
Jan 13 14:52:43 raspberrypi kolibri-server[8848]: start-stop-daemon: matching on world-writable pidfile /var/run/kolibri-server/uwsgi.pid is insecure
Jan 13 14:52:55 raspberrypi systemd[1]: kolibri-server.service: Stopping timed out. Terminating.
Jan 13 14:52:55 raspberrypi systemd[1]: kolibri-server.service: Control process exited, code=killed, status=15/TERM
Jan 13 14:52:55 raspberrypi systemd[1]: kolibri-server.service: Killing process 8988 (systemctl) with signal SIGKILL.
Jan 13 14:52:55 raspberrypi systemd[1]: kolibri-server.service: Failed with result 'timeout'.
Jan 13 14:52:55 raspberrypi systemd[1]: Failed to start A high performance web server setup for Kolibri.
dpkg: error processing package kolibri-server (--configure):
 installed kolibri-server package post-installation script subprocess returned error exit status 1
Errors were encountered while processing:
 kolibri-server
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

### PR DESCRIPTION
**Pull request by jredrejo**
After investigating deeply #51 , there were several reasons for the service to timeout:
- Using service to start/stop/restart kolibri was delaying everything because systemd configuration for kolibri has a `TimeoutStartSec=infinity` setup
- Sometimes the pid and sockets for uwsgi were not correctly deleted
- kolibri-server systemd was more complex than needed

This PR:
- Uses the same commands the kolibri daemon uses to start it, but without depending on systemctl
- Optimizes uwsgi configuration
- Optimizes systemd configuration
- Ensures any leftover socket and pid for uwsgi is removed before starting

Close #51 
Cancels #50

### CODE REVIEW

File: uwsgi.ini [hunk 1]
```diff
@@ -25,10 +25,24 @@ vacuum = True       # Try to remove all of the generated files/sockets (UNIX soc
...
+
+disable-logging = true               ; Disable built-in logging 
+log-4xx = true                       ; but log 4xx's anyway
+log-5xx = true                       ; and 5xx's
+
+py-callos-afterfork = true           ; allow workers to trap signals
+
+max-requests = 1000                  ; Restart workers after this many requests
+max-worker-lifetime = 3600           ; Restart workers after this many seconds
+worker-reload-mercy = 60             ; How long to wait before forcefully killing workers
```
Comment (indirectlylit) [inline uwsgi.ini]: super trivial, but just flagging that we're mixing comment delimiters in the same file (`;` vs `#`)

**Review by indirectlylit — COMMENTED** [review]
Are there any consequences to not using `systemd`? What sorts of functionality comes with the complexity that we'd be bypassing?

File: uwsgi.ini (hunk 1 above)
Comment (jredrejo) [reply uwsgi.ini]: No so trivial, thanks. Fixed

**Comment by jredrejo** [comment]
> Are there any consequences to not using `systemd`? What sorts of functionality comes with the complexity that we'd be bypassing?

We keep using systemd for `kolibri-server`, it's needed for any standard Debian based installation.
However, inside the script running `kolibri-server `the code was using systemd to start the `kolibri` service. Now the `kolibri-server` script is using the same commands the `kolibri` service uses but skipping the systemd timeouts that are set in the `kolibri` package.
The only consequence I can think of is that, if in the future, the `kolibri` package changes the way of executing kolibri commands, `kolibri-server` will need to keep it in sync. That should not be a problem with the current relation between these two packages and their maintainers.

P.S sorry for the excesive repetition of `kolibri` and `kolibri-server` names but I haven't found a better way to explain it.

**Comment by benjaoming** [comment]
How do we handle `kolibri` package's command outside of this? By starting/stopping the system service (sysvinit or systemd) from `kolibri-server`, we ensure that there are no conflicts, if the service is already started, it will be ignored.

But by directly starting Kolibri, we no longer have the same controls that an individual system service provides, which is part of all the logic that the `kolibri` debian package is responsible for.

I feel like we are going to maintain the same patterns in two packages now?

> Using service to start/stop/restart kolibri was delaying everything because systemd configuration for kolibri has a TimeoutStartSec=infinity setup

Didn't it have this for a good reason?

> Sometimes the pid and sockets for uwsgi were not correctly deleted

How is this related to a separate system service that just runs Kolibri's background jobs/scheduler?

**Comment by benjaoming** [comment]
> kolibri-server systemd was more complex than needed

Have you gone over the option of dependencies between the two system services? Such that we create a systemd/sysvinit config for `kolibri-server` that depends on the `kolibri` system service? `systemd` does these things, not sure about sysvinit, Upstart and such. Maybe Upstart is already so deprecated it doesn't matter.

File: debian/kolibri-server.service [hunk 2]
```diff
@@ -10,12 +10,7 @@ Wants=network-online.target
 Type=forking
 ExecStart=/etc/init.d/kolibri-server start
 ExecStop=/etc/init.d/kolibri-server stop
-TimeoutStartSec=infinity
```
Comment (benjaoming) [inline debian/kolibri-server.service:13]: Was it also the intention to remove this?

File: uwsgi.ini [hunk 3]
```diff
@@ -25,10 +25,24 @@ vacuum = True       # Try to remove all of the generated files/sockets (UNIX soc
 module =kolibri.deployment.default.wsgi:application
 buffer-size = 8192  # needed to support long requests blocks (in lessons, for example)
 wsgi-disable-file-wrapper = true  # needed to fix uwsgi bug https://github.com/unbit/uwsgi/issues/1126
+single-interpreter = true
+die-on-term = true                   # Shutdown when receiving SIGTERM (default is respawn)
+need-app = true
+
+disable-logging = true               # Disable built-in logging
+log-4xx = true                       # but log 4xx's anyway
+log-5xx = true                       # and 5xx's
```
Comment (benjaoming) [inline uwsgi.ini:34]: This is because we don't have log rotation?

**Review by benjaoming — APPROVED** [review]
Noting to everything that I just mentioned in above comments: The `kolibri` debian package has an extremely simple system service, and nothing here seems to be a regression to me. I'm more concerned/speculating about future potential improvements in the `kolibri` package's system script having to be duplicated here.

File: uwsgi.ini (hunk 3 above)
Comment (jredrejo) [reply uwsgi.ini:34]: No, it's to avoid duplication. The same info is available in the nginx logs, excepting the python errors.

**Comment by jredrejo** [comment]
> Noting to everything that I just mentioned in above comments: The `kolibri` debian package has an extremely simple system service, and nothing here seems to be a regression to me. I'm more concerned/speculating about future potential improvements in the `kolibri` package's system script having to be duplicated here.

I agree that this could take us to duplicate efforts and I had spent a long time trying to find a workaround but, really, this is needed: using a systemctl call from a systemctl script was causing timeouts. Actually I had not seen any other example of this tactic and I guess there are good reasons for it. Removing the call and using exactly the same commands the kolibri daemon uses removed all the timeouts and the server starts in less than 5 seconds in a Raspberry. Before, it could not start in 90 seconds.

In any case, all the people involved in the kolibri `package` are also somehow involved in the `kolibri-server` package. I trust we are not going to lose any sync between future improvements.

File: debian/kolibri-server.service (hunk 2 above)
Comment (jredrejo) [reply debian/kolibri-server.service:13]: Yes, after removing the `service` call in the service script this was unneeded. Default system values are more than enough.

File: debian/kolibri-server.init [hunk 4]
```diff
@@ -94,19 +93,20 @@ do_check_nginx()
...
+    su $KOLIBRI_USER -c "$KOLIBRI_COMMAND restart &" 
     service nginx start
   else
     service nginx reload
   fi
 }
 
 do_stop() {
-  service kolibri stop
+  su $KOLIBRI_USER -c "$KOLIBRI_COMMAND stop &"
```
Comment (benjaoming) [inline debian/kolibri-server.init]: Is this risky? For instance, if a system is shutting down, could it cause the process to be killed prematurely?

File: debian/kolibri-server.service [hunk 5]
```diff
@@ -10,12 +10,7 @@ Wants=network-online.target
...
-TimeoutStopSec=10
```
Comment (benjaoming) [inline debian/kolibri-server.service:14]: Wasn't this timeout supposed to have guaranteed stopping within 10 seconds?

Comment (benjaoming) [reply debian/kolibri-server.service:14]: Anyways, I agree that we can use default system values and see how that goes... if Stop takes over 10 seconds, I want to know and not have released forced killing of the system process in-the-wild :)

**Comment by benjaoming** [comment]
Released in kolibri-proposed as kolibri-server_0.3.7\~beta1-0ubuntu1

**Comment by benjaoming** [comment]
All of this might as well be rooted in https://github.com/learningequality/kolibri-installer-debian/issues/90 since `kolibri-server.service` starts/stops the normal `kolibri.service`.

**Comment by benjaoming** [comment]
Released in kolibri-proposed as kolibri-server_0.3.7~beta1-0ubuntu2

**Comment by benjaoming** [comment]
An updated Kolibri 0.13.2~beta1 is also available in kolibri-proposed, so all the fixes can be tested together.

File: debian/kolibri-server.init (hunk 4 above)
Comment (benjaoming) [reply debian/kolibri-server.init]: Assuming this is risky: https://github.com/learningequality/kolibri-server/pull/28

**Comment by benjaoming** [comment]
Instead of testing this individually, I'm releasing as 0.3.8 to get #53 tested thoroughly as well: https://github.com/learningequality/kolibri-server/pull/61

File: debian/kolibri-server.init [hunk 6]
```diff
@@ -94,20 +104,18 @@ do_check_nginx()
...
+    $SU_COMMAND $KOLIBRI_USER -c "$KOLIBRI_COMMAND restart"
     service nginx start
   else
     service nginx reload
   fi
 }
 
 do_stop() {
-  service kolibri stop
+  $SU_COMMAND $KOLIBRI_USER -c "$KOLIBRI_COMMAND stop"
```
Comment (jredrejo) [inline debian/kolibri-server.init:115]: Why did you remove the `&` at the end of the command? locking causes important delays when restarting the associate processes. Same thing with the restart command.
It causes that, for example, restarting or halting the server takes an unneeded time while waiting for kolibri to shutdown the zeroconf daemons, what is unneeded both restarting or stopping.
Did you see any issue by adding it?

File: debian/kolibri-server.init [hunk 7]
```diff
@@ -88,9 +88,10 @@ do_stop_uwsgi()
   #   other if a failure occurred
   start-stop-daemon --stop --quiet --retry=TERM/30/KILL/5 --pidfile $PIDFILE_UWSGI --name uwsgi
   RETVAL="$?"
-  [ "$RETVAL" = 2 ] && return 2
   # Many daemons don't delete their pidfiles when they exit.
   rm -f $PIDFILE_UWSGI
+  rm -f /tmp/kolibri_uwsgi.sock
```
Comment (jredrejo) [inline debian/kolibri-server.init:93]: I don't understand why these lines have been moved here, removing the uwsgi lock before uwsgi stops. What's the reason?

File: debian/kolibri-server.init (hunk 6 above)
Comment (benjaoming) [reply debian/kolibri-server.init:115]: Commented on that further up: https://github.com/learningequality/kolibri-server/pull/54#discussion_r393977573

Forking `kolibri stop` to the background and then restarting it simultaneously seems to be asking for unexpected behavior? I don't think it's worth it if -- theoretically -- it means more malformed databases or broken upgrades (restart is a part of upgrading, I think).

It could also basically just make `restart` broken, in cases where kolibri refuses to start because another instance still hasn't stopped?

Comment (benjaoming) [reply debian/kolibri-server.init:115]: I understand it can make stopping the service faster.. but https://github.com/learningequality/kolibri-installer-debian/issues/90 has a promising fix for the perhaps most annoying part :)

File: debian/kolibri-server.init (hunk 7 above)
Comment (benjaoming) [reply debian/kolibri-server.init:93]: UWSGI is stopped further up, I moved the whole logic into `do_stop_uwsgi` -- before it had to be run twice because it was doing `[ "$RETVAL" = 2 ] && return 2` firstly.

File: debian/kolibri-server.init (hunk 4 above)
Comment (jredrejo) [reply debian/kolibri-server.init]: no, it's not risky (at least not in my tests nor my logic). For #28 the problem is kolibri not stopping. This should not happen here becaue the command is executed (in the background, but executed).
And when the machine is restarted, waiting for zeroconf to finish its shutdown is not needed.

I tested it in different machines (including RPi) and worked properly but don't take it as a warranty, if you can test it more, much better. The rationale for the `&` was to avoid restarting a machine is locked for more than one minute waiting for zeroconf.  I think the tests are worth if we can include this change.

File: debian/kolibri-server.service (hunk 5 above)
Comment (jredrejo) [reply debian/kolibri-server.service:14]: Please, test it , four eyes will see more than two. This is one of the changes I tested in:
-  RPi 4
-  a Virtualbox vm
- my own laptop (good cpu with 16 Gb of RAM)

If you can add more scenarios, much better. 
Also I took a look at several Debian packages managing daemons and none of them were using this Timeout options. It seems the defaults usually work

File: debian/kolibri-server.init (hunk 6 above)
Comment (jredrejo) [reply debian/kolibri-server.init:115]: If https://github.com/learningequality/kolibri-installer-debian/issues/90 works we can skip this change. If not, I'd keep it at least for the `stop` part. That's what affects halting a machine and zeroconf does not matter at that moment. In the `restart` part we would have to live with the delay if we see it produces issues.

File: debian/kolibri-server.init (hunk 7 above)
Comment (jredrejo) [reply debian/kolibri-server.init:93]: gotcha, thanks!

File: debian/kolibri-server.init (hunk 6 above)
Comment (benjaoming) [reply debian/kolibri-server.init:115]: I would feel a lot better about the `&` addition if it went through a stress test.. like a very fast system running `systemctl restart kolibri` 100 times. We should do that some day in any case, so once we get around to that, we can see if stopping can be made more concurrent?

Adding a follow-up issue: https://github.com/learningequality/kolibri-server/issues/62

Comment (jredrejo) [reply debian/kolibri-server.init:115]: I understand your concern when using `restart` (and I agree to remove the `&` until we test it more) but I can't find a reason for problems when using `stop`, why do you think the service should be waiting for kolibri to stop instead of launching it in the background?

Comment (benjaoming) [reply debian/kolibri-server.init:115]: Sorry I have been grasping this sporadically -- forget everything about `restart`, what worried me originally was breaking stuff during upgrades.

This would hold true for the `kolibri` package (but we're not adding `&` there, so it's not relevant), but I can't find out what the defined behavior is for `kolibri-server` package?

Here's what I feared could happen:

```
sudo apt upgrade

kolibri.preinst (debian script)
=> Invokes systemctl stop kolibri / service kolibri stop

   Time 0: "kolibri stop &"  <= forks into background
   Time 0: "start-stop-daemon uwsgi" <= blocks
   Time 1: UWSGI is now stopped

kolibri.postinst
=> Invokes systemctl start kolibri / service kolibri start

   Time 2: "kolibri start"
   Time 3: "kolibri stop" finishes # called at Time 0
```

If apt/dpkg issues a service `restart` after the upgrade, then it's all safe. If it stops the service before the upgrade and starts it again after, I would consider it unsafe.
